### ISSUE TITLE
Local .terraform cache prevents S3 backend configuration changes to take effect

### ISSUE THREAD

**Issue by uhef** [hashicorp/terraform#22488]
### Terraform Version

```
Terraform v0.11.14
```

### Terraform Configuration Files

First run (without `role_arn`):
```hcl
terraform {
  backend "s3" {
    bucket               = "statebucket.organization.org"
    key                    = "path/to/state/terraform.tfstate"
    region                = "eu-west-1"
    encrypt              = true
    dynamodb_table       = "terraform-lock"
  }
}
```

Second run (with `role_arn`):
```
terraform {
  backend "s3" {
    bucket               = "statebucket.organization.org"
    key                    = "path/to/state/terraform.tfstate"
    region                = "eu-west-1"
    encrypt              = true
    dynamodb_table       = "terraform-lock"
    role_arn             = "arn:aws:iam::1234567890:role/Administrator"
  }
}
```

### Debug Output

Obfuscated debug output from the first run (before explicitly specifying `role_arn`): https://gist.github.com/uhef/3ec922c765c1cb1affbe970d70437d33

Obfuscated debug output from the second run (after explicitly specifying `role_arn`): 
https://gist.github.com/uhef/02cdc6be7e50a77012c6f41c180e7199

Obfuscated debug output from the third run (after running `rm -rf .terraform`): 
https://gist.github.com/uhef/ecc128f1907e2d151acf1ef2383efcfc

### Expected Behavior
Once a role assumption role that has required permissions to access S3 state bucket in `role_arn` property of S3 backend had been defined, Terraform should have been able to initialize the project successfully without any extra actions from the user, such as clearing the local Terraform cache.

### Actual Behavior
It requires an extra clearing of local Terraform cache before the `terraform init` succeeds. This can be done by `rm -rf .terraform`.

### Steps to Reproduce
1. Create a new Terraform configuration using S3 backend.
1. Run `terraform init` as a user that does not directly have permissions to execute s3/ListObjects on the target S3 bucket. Observe that the initialization fails (as it should)
1. Add `role_arn` property to the S3 backend configuration to assume role that does have the s3/ListObjects permissions.
1. Run `terraform init`. Observe that the initialization still fails (although it shouldn't). Also notice from logs that Terraform does two `GetCallerIdentity` calls. Once with the assumed role and once with the original user from step 2.
1. Remove local cache by running `rm -rf .terraform`.
1. Rerun `terraform init`. Notice that now it succeeds.

**Comment by mildwonkey** [comment]
Hi @uhef , thank you for reporting this behavior!

I've labeled this as a `backend/s3` bug, but this is expected behavior from all backends. If Terraform is unable to authenticate to the backend provider on the first init, subsequent `init`s will fail because Terraform is using the original configuration to pull the state _before_ it applies the new configuration. 

When the original `init` fails, you can run [`terraform init -reconfigure`](https://www.terraform.io/docs/commands/init.html#backend-initialization) to tell Terraform to ignore the original (mis)configuration instead of deleting the `.terraform` directory.

To improve this situation, we could append a message to the output when the backend init fails that reminds the user of the `init -reconfigure` option

**Comment by jbardin** [comment]
Closed via #28718
